### PR TITLE
Adds sqlalchemy pool values for nova.

### PR DESCRIPTION
**Pull request by dguitarbite**
Add sqlalchemy pool values to avoid hardcoding DB pool values for nova.

### CODE REVIEW

File: chef/cookbooks/nova/templates/default/nova.conf.erb [hunk 1]
```diff
@@ -2748,7 +2748,7 @@ connection=<%= @database_connection %>
 # (integer value)
 # Deprecated group/name - [DEFAULT]/sql_max_pool_size
 # Deprecated group/name - [DATABASE]/sql_max_pool_size
-#max_pool_size=<None>
+max_pool_size=<%= node[:nova][:db][:max_pool_size] %>
```
Comment (toabctl) [inline chef/cookbooks/nova/templates/default/nova.conf.erb]: please leave the default as comment. Make config template updates for later releases easier

File: chef/cookbooks/nova/templates/default/nova.conf.erb [hunk 2]
```diff
@@ -2766,7 +2766,7 @@ connection=<%= @database_connection %>
 # (integer value)
 # Deprecated group/name - [DEFAULT]/sql_max_overflow
 # Deprecated group/name - [DATABASE]/sqlalchemy_max_overflow
-#max_overflow=<None>
+max_overflow=<%= node[:nova][:db][:max_overflow] %>
```
Comment (toabctl) [inline chef/cookbooks/nova/templates/default/nova.conf.erb]: same here

File: chef/cookbooks/nova/templates/default/nova.conf.erb [hunk 3]
```diff
@@ -2781,7 +2781,7 @@ connection=<%= @database_connection %>
 # If set, use this value for pool_timeout with sqlalchemy
 # (integer value)
 # Deprecated group/name - [DATABASE]/sqlalchemy_pool_timeout
-#pool_timeout=<None>
+pool_timeout=<%= node[:nova][:db][:pool_timeout] %>
```
Comment (toabctl) [inline chef/cookbooks/nova/templates/default/nova.conf.erb]: same here

File: chef/data_bags/crowbar/bc-template-nova.json [hunk 4]
```diff
@@ -80,7 +83,7 @@
     "nova": {
       "crowbar-revision": 0,
       "crowbar-applied": false,
-      "schema-revision": 33,
+      "schema-revision": 34,
```
Comment (vuntz) [inline chef/data_bags/crowbar/bc-template-nova.json]: Asking out loud: did we bump to 30 after branching for Cloud 5, or should be bump to 40?

Comment (dguitarbite) [inline chef/data_bags/crowbar/bc-template-nova.json]: @vuntz IMHO I do not know much about the version numbering, just incremented it by 1.

**Comment by toabctl** [comment]
please remove the master2 merge commit

File: chef/data_bags/crowbar/bc-template-nova.json (hunk 4 above)
Comment (tboerger) [inline chef/data_bags/crowbar/bc-template-nova.json]: Just look for the version in Cloud 5, if it is 2x your change is fine, if not we should raise it.

**Comment by mjura** [comment]
Please rebase this PR

**Comment by toabctl** [comment]
@dguitarbite please address my comments about the comments in the template.

**Comment by dguitarbite** [comment]
@toabctl  I think I did update the comments, let me check if it got lost in the rebase.

File: chef/data_bags/crowbar/bc-template-nova.json (hunk 4 above)
Comment (dguitarbite) [inline chef/data_bags/crowbar/bc-template-nova.json]: Just checked release/tex/master the version is 2x. Keeping the numbers same since its 3.x for cloud 6.

File: chef/cookbooks/nova/templates/default/nova.conf.erb [hunk 5]
```diff
@@ -2748,7 +2748,7 @@ connection=<%= @database_connection %>
 # (integer value)
 # Deprecated group/name - [DEFAULT]/sql_max_pool_size
 # Deprecated group/name - [DATABASE]/sql_max_pool_size
-#max_pool_size=<None>
+#max_pool_size=<%= node[:nova][:db][:max_pool_size] %>
```
Comment (toabctl) [inline chef/cookbooks/nova/templates/default/nova.conf.erb]: So what I would like to see is:
```
#max_pool_size=<None>
max_pool_size=<%= node[:nova][:db][:max_pool_size] %>
```

File: chef/data_bags/crowbar/migrate/nova/034_add_sqlalchemy_pool_values.rb [hunk 6]
```diff
@@ -0,0 +1,13 @@
+def upgrade(ta, td, a, d)
+  a["db"]["max_pool_size"] = ta["db"]["max_pool_size"]
```
Comment (toabctl) [inline chef/data_bags/crowbar/migrate/nova/034_add_sqlalchemy_pool_values.rb]: If we are going to backport this, you want to check for already existing variables

Comment (dguitarbite) [inline chef/data_bags/crowbar/migrate/nova/034_add_sqlalchemy_pool_values.rb]: @toabctl Yes, I removed then thinking its not required, Ill check if we need  to backport it.

File: chef/data_bags/crowbar/migrate/nova/034_add_sqlalchemy_pool_values.rb [hunk 7]
```diff
@@ -0,0 +1,25 @@
+def upgrade(ta, td, a, d)
+  unless ta["db"].has_key? "max_pool_size"
```
Comment (toabctl) [inline chef/data_bags/crowbar/migrate/nova/034_add_sqlalchemy_pool_values.rb]: Must be ```unless a["db"].has_key```

File: chef/data_bags/crowbar/migrate/nova/034_add_sqlalchemy_pool_values.rb [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+    a["db"]["max_pool_size"] = ta["db"]["max_pool_size"]
+  end
+  unless ta["db"].has_key? "max_overflow"
```
Comment (toabctl) [inline chef/data_bags/crowbar/migrate/nova/034_add_sqlalchemy_pool_values.rb]: same here

File: chef/data_bags/crowbar/migrate/nova/034_add_sqlalchemy_pool_values.rb [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+    a["db"]["max_overflow"] = ta["db"]["max_overflow"]
+  end
+  unless ta["db"].has_key? "pool_timeout"
```
Comment (toabctl) [inline chef/data_bags/crowbar/migrate/nova/034_add_sqlalchemy_pool_values.rb]: same here

File: chef/data_bags/crowbar/migrate/nova/034_add_sqlalchemy_pool_values.rb [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+    a["db"]["pool_timeout"] = ta["db"]["pool_timeout"]
+  end
+  return a, d
+end
+
+def downgrade(ta, td, a, d)
+  unless ta["db"].has_key? "max_pool_size"
```
Comment (tboerger) [inline chef/data_bags/crowbar/migrate/nova/034_add_sqlalchemy_pool_values.rb]: IMHO that doesn't make sense

Comment (dirkmueller) [inline chef/data_bags/crowbar/migrate/nova/034_add_sqlalchemy_pool_values.rb]: It makes sense to me.

File: chef/data_bags/crowbar/migrate/nova/034_add_sqlalchemy_pool_values.rb [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+    a["db"]["max_pool_size"].delete
```
Comment (dirkmueller) [inline chef/data_bags/crowbar/migrate/nova/034_add_sqlalchemy_pool_values.rb]: did you try that ? that gives a syntax error for me:

ArgumentError: wrong number of arguments (0 for 1+

File: chef/data_bags/crowbar/migrate/nova/034_add_sqlalchemy_pool_values.rb (hunk 10 above)
Comment (tboerger) [inline chef/data_bags/crowbar/migrate/nova/034_add_sqlalchemy_pool_values.rb]: So "if ta["db"] has not the key max_pool_size delete the key max_pool_size" makes sense to you?

File: chef/data_bags/crowbar/migrate/nova/034_add_sqlalchemy_pool_values.rb (hunk 11 above)
Comment (tboerger) [inline chef/data_bags/crowbar/migrate/nova/034_add_sqlalchemy_pool_values.rb]: Already told that, he @dguitarbite said you suggested that :)
It have to be `a["db"].delete "max_pool_size"`

Comment (dguitarbite) [inline chef/data_bags/crowbar/migrate/nova/034_add_sqlalchemy_pool_values.rb]: Yes, I have updated the code. I am waiting for my local setup to run so that I can test this.

**Comment by vuntz** [comment]
+1

File: chef/data_bags/crowbar/migrate/nova/034_add_sqlalchemy_pool_values.rb [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+    a["db"]["max_overflow"] = ta["db"]["max_overflow"]
+  end
+  unless a["db"].has_key? "pool_timeout"
+    a["db"]["pool_timeout"] = ta["db"]["pool_timeout"]
+  end
+  return a, d
+end
+
+def downgrade(ta, td, a, d)
+  if a["db"].has_key? "max_pool_size"
```
Comment (dirkmueller) [inline chef/data_bags/crowbar/migrate/nova/034_add_sqlalchemy_pool_values.rb]: it should be

unless ta["db"].has_key? "max_pool_size"

File: chef/data_bags/crowbar/migrate/nova/034_add_sqlalchemy_pool_values.rb [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+    a["db"].delete "max_pool_size"
+  end
+  if a["db"].has_key? "max_overflow"
```
Comment (dirkmueller) [inline chef/data_bags/crowbar/migrate/nova/034_add_sqlalchemy_pool_values.rb]: same here

File: chef/data_bags/crowbar/migrate/nova/034_add_sqlalchemy_pool_values.rb [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+    a["db"].delete "max_overflow"
+  end
+  if a["db"].has_key? "pool_timeout"
```
Comment (dirkmueller) [inline chef/data_bags/crowbar/migrate/nova/034_add_sqlalchemy_pool_values.rb]: same here

File: chef/data_bags/crowbar/template-nova.schema [hunk 15]
```diff
@@ -8,8 +8,8 @@
       "type": "map",
       "required": true,
       "mapping": {
-        "nova": { 
-          "type": "map", 
+        "nova": {
+          "type": "map",
```
Comment (dirkmueller) [inline chef/data_bags/crowbar/template-nova.schema]: unrelated change? perhaps in a 2nd commit

**Comment by dirkmueller** [comment]
please rebase, not merge

File: chef/data_bags/crowbar/template-nova.schema (hunk 15 above)
Comment (dguitarbite) [inline chef/data_bags/crowbar/template-nova.schema]: Mostly its my vim auto-cleaning trailing whitespaces. Ill remove it from this commit.

File: chef/data_bags/crowbar/migrate/nova/035_add_sqlalchemy_pool_values.rb [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+    a["db"]["max_overflow"] = ta["db"]["max_overflow"]
+  end
+  unless a["db"].has_key? "pool_timeout"
+    a["db"]["pool_timeout"] = ta["db"]["pool_timeout"]
+  end
+  return a, d
+end
+
+def downgrade(ta, td, a, d)
+  unless ta["db"].has_key? "max_pool_size"
```
Comment (tboerger) [inline chef/data_bags/crowbar/migrate/nova/035_add_sqlalchemy_pool_values.rb]: That's still crap, as commented before!

**Comment by tboerger** [comment]
-0.5
The downgrade method is still totally wrong.

File: chef/data_bags/crowbar/migrate/nova/035_add_sqlalchemy_pool_values.rb (hunk 16 above)
Comment (dguitarbite) [inline chef/data_bags/crowbar/migrate/nova/035_add_sqlalchemy_pool_values.rb]: Is the problem with unless? I think 'if ta....' would make more sense here.

File: chef/data_bags/crowbar/template-nova.schema [hunk 17]
```diff
@@ -47,14 +47,17 @@
...
+                "max_overflow": { "type": "int", "required": true },
+                "pool_timeout": { "type": "int", "required": true }
               }
             },
             "rbd": {
               "type": "map", "required": false, "mapping": {
                 "user": { "type": "str", "required": true },
                 "secret_uuid": { "type": "str", "required": true }
-              }  
+              }
```
Comment (dirkmueller) [inline chef/data_bags/crowbar/template-nova.schema]: move this to the whitespace commit..

File: chef/data_bags/crowbar/migrate/nova/035_add_sqlalchemy_pool_values.rb (hunk 16 above)
Comment (tboerger) [inline chef/data_bags/crowbar/migrate/nova/035_add_sqlalchemy_pool_values.rb]: As i commented before

File: chef/data_bags/crowbar/template-nova.schema [hunk 18]
```diff
@@ -8,8 +8,8 @@
       "type": "map",
       "required": true,
       "mapping": {
-        "nova": { 
-          "type": "map", 
+        "nova": {
```
Comment (dirkmueller) [inline chef/data_bags/crowbar/template-nova.schema]: move this to the whitespace commit?

File: chef/data_bags/crowbar/migrate/nova/034_add_sqlalchemy_pool_values.rb (hunk 10 above)
Comment (dirkmueller) [inline chef/data_bags/crowbar/migrate/nova/034_add_sqlalchemy_pool_values.rb]: Yes. it means "if the old schema doesn't *require* the key, then we remove the key from the attributes". it is doing a proper schema downgrade even if the change got backported to an older code release (by which the schema would then suddenly require this key to exist). 

this is really standard code that we use in hundreds or cases already, I really don't get what would be different here..

**Comment by tboerger** [comment]
Forget my comment and my -0.5

File: chef/data_bags/crowbar/template-nova.json [hunk 19]
```diff
@@ -30,7 +30,10 @@
       "db": {
         "password": "",
         "user": "nova",
-        "database": "nova"
+        "database": "nova",
+        "max_pool_size": 30,
```
Comment (dirkmueller) [inline chef/data_bags/crowbar/template-nova.json]: where do those defaults come from? it looks like the upstream default is "None" and now we're hardcoding something..

Comment (dguitarbite) [inline chef/data_bags/crowbar/template-nova.json]: Yes, upstream defaults to none. I took the most commonly used values. I can reset them to none if you think that would be a better idea.

File: chef/cookbooks/nova/attributes/default.rb [hunk 20]
```diff
@@ -28,6 +28,11 @@
 default[:nova][:db][:user] = "nova"
 default[:nova][:db][:database] = "nova"
 
+# Sqlalchemy parameters
```
Comment (rsalevsky) [inline chef/cookbooks/nova/attributes/default.rb]: SQLAlchemy

**Comment by tboerger** [comment]
I have no idea if the default values are ok or usable but LGTM codewise.

File: chef/data_bags/crowbar/template-nova.json (hunk 19 above)
Comment (dirkmueller) [inline chef/data_bags/crowbar/template-nova.json]: the worse thing is that we don't seem to have  a method to change the values back to "<None>" now it seems

Comment (dguitarbite) [inline chef/data_bags/crowbar/template-nova.json]: From the previous patch I got the base values, then looked online on different discussions and forums.

I will try to take another look at getting something similar to "none".

**Comment by toabctl** [comment]
@dguitarbite there is now only a whitespace commit?

**Comment by dguitarbite** [comment]
@toabctl Thanks for noticing, Ill update this.

File: chef/data_bags/crowbar/template-nova.json (hunk 19 above)
Comment (toabctl) [inline chef/data_bags/crowbar/template-nova.json]: > the worse thing is that we don't seem to have a method to change the values back to "" now it seems.

We could make the parameter optional in the barclamp and then set it only if something is given. That method is used in the manila barclamp. see for example:
- https://github.com/crowbar/crowbar-openstack/blob/master/chef/cookbooks/manila/templates/default/manila.conf.erb#L1866
- https://github.com/crowbar/crowbar-openstack/blob/master/chef/data_bags/crowbar/template-manila.schema#L61


The downside is that not all available keys are listed in a proposal.

Comment (toabctl) [inline chef/data_bags/crowbar/template-nova.json]: @dirkmueller @dguitarbite opinions about making the parameters optional?

**Comment by mjura** [comment]
Why don't you allow also setup `min_pool_size=` option as well?

**Comment by dguitarbite** [comment]
@mjura Yes I will add `min_pool_size=` option if it is relevant to the db settings.

File: chef/data_bags/crowbar/template-nova.json (hunk 19 above)
Comment (dguitarbite) [inline chef/data_bags/crowbar/template-nova.json]: @toabctl Sounds like the best way for this PR. There is no way to know what default values to use. Thanks for the suggestion.

File: chef/cookbooks/nova/attributes/default.rb [hunk 21]
```diff
@@ -28,6 +28,12 @@
 default[:nova][:db][:user] = "nova"
 default[:nova][:db][:database] = "nova"
 
+# SQLAlchemy parameters
+default[:nova][:db][:max_pool_size] = -1
+default[:nova][:db][:max_overflow] = -1
+default[:nova][:db][:pool_timeout] = -1
+default[:nova][:db][:min_pool_size] = -1
```
Comment (toabctl) [inline chef/cookbooks/nova/attributes/default.rb]: not sure, but maybe -1 are valid values? I think you could use nil if you don't want to set the values.

File: chef/cookbooks/nova/templates/default/nova.conf.erb [hunk 22]
```diff
@@ -2748,13 +2748,18 @@ connection=<%= @database_connection %>
 # Deprecated group/name - [DEFAULT]/sql_min_pool_size
 # Deprecated group/name - [DATABASE]/sql_min_pool_size
 #min_pool_size=1
+<% if node[:nova][:db][:min_pool_size] < 0 -%>
```
Comment (toabctl) [inline chef/cookbooks/nova/templates/default/nova.conf.erb]: Wrong logic. Use ```unless```

File: chef/cookbooks/nova/templates/default/nova.conf.erb [hunk 23]
```diff
@@ -2748,13 +2748,18 @@ connection=<%= @database_connection %>
...
+min_pool_size=<%= node[:nova][:db][:min_pool_size] -%>
+<% end -%>
 
 # Maximum number of SQL connections to keep open in a pool
 # (integer value)
 # Deprecated group/name - [DEFAULT]/sql_max_pool_size
 # Deprecated group/name - [DATABASE]/sql_max_pool_size
 #max_pool_size=<None>
-
+<% if node[:nova][:db].[:max_pool_size] -%>
```
Comment (toabctl) [inline chef/cookbooks/nova/templates/default/nova.conf.erb]: syntax error. Have you ever tested that?

**Comment by toabctl** [comment]
Superseeded by https://github.com/crowbar/crowbar-openstack/pull/126
